libp2p/protocols/connectivity/autonatv2/client.nim

```diff
@@ -0,0 +1,117 @@
+# Nim-LibP2P
+# Copyright (c) 2025 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.push raises: [].}
+
+import results
+import chronos, chronicles
+import
+  ../../protocol,
+  ../../../switch,
+  ../../../multiaddress,
+  ../../../multicodec,
+  ../../../peerid,
+  ../../../protobuf/minprotobuf,
+  ./types
+
+logScope:
+  topics = "libp2p autonat v2 client"
+
+const
+  MaxAcceptedDialDataRequest* = 100 * 1024 # 100 KB
+  MaxDialDataResponsePayload* = 1024
+
+type AutonatV2Client* = ref object of RootObj
+  rng: ref HmacDrbgContext
+  nonces: Table[PeerId, uint64]
+
+proc new*(T: typedesc[AutonatV2Client], rng: ref HmacDrbgContext, switch: Switch): T =
+  let client = T(rng: rng, nonces: initTable[PeerId, uint64]())
+  client
+
+proc handleDialDataRequest*(
+    conn: Connection, req: DialDataRequest
+): Future[Opt[DialResponse]] {.
+    async: (raises: [CancelledError, AutonatV2Error, LPStreamError])
+.} =
+  debug "handleDialDataRequest"
+  defer:
+    await conn.close()
+
+  if req.numBytes > MaxAcceptedDialDataRequest:
+    error "Rejecting DialDataRequest: numBytes is greater than the maximum",
+      numBytes = req.numBytes, maxAcceptedNumBytes = MaxAcceptedDialDataRequest
+    return Opt.none(DialResponse)
+
+  # send required data
+  var msg = AutonatV2Msg(
+    msgType: MsgType.DialDataResponse,
+    dialDataResp: DialDataResponse(data: newSeq[byte](MaxDialDataResponsePayload)),
+  )
+  let messagesToSend =
+    (req.numBytes + MaxDialDataResponsePayload - 1) div MaxDialDataResponsePayload
+  for i in 0 ..< messagesToSend:
+    await conn.send(msg)
+    debug "sending dialdataresponse", i = i, messagesToSend = messagesToSend
+
+  # get DialResponse
+  msg = AutonatV2Msg.decode(initProtoBuffer(await conn.readLp(1024))).valueOr:
+    raise newException(AutonatV2Error, "Unable to decode message")
```

Comment:
> Unable to decode message

okay, but will developer know what message is that based on logs?

Reply:
Hummm indeed. However the problem is that message could be total garbage, or only partial garbage. In both cases decoding would fail. Should we print the raw bytes in a debug log?